bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/connector/EpsonProjectorSerialConnector.java

```diff
@@ -0,0 +1,214 @@
+import org.openhab.core.io.transport.serial.SerialPortIdentifier;
+import org.openhab.core.io.transport.serial.SerialPortManager;
+import org.openhab.core.io.transport.serial.UnsupportedCommOperationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Connector for serial port communication.
+ *
+ * @author Pauli Anttila - Initial contribution
+ * @author Michael Lobstein - Improvements for OH3
+ */
+@NonNullByDefault
+public class EpsonProjectorSerialConnector implements EpsonProjectorConnector, SerialPortEventListener {
+
+    private final Logger logger = LoggerFactory.getLogger(EpsonProjectorSerialConnector.class);
+    private final String serialPortName;
+    private final SerialPortManager serialPortManager;
+
+    private @Nullable InputStream in = null;
+    private @Nullable OutputStream out = null;
+    private @Nullable SerialPort serialPort = null;
+
+    public EpsonProjectorSerialConnector(SerialPortManager serialPortManager, String serialPort) {
+        this.serialPortManager = serialPortManager;
+        this.serialPortName = serialPort;
+    }
+
+    @Override
+    public void connect() throws EpsonProjectorException {
+        try {
+            logger.debug("Open connection to serial port '{}'", serialPortName);
+
+            SerialPortIdentifier serialPortIdentifier = serialPortManager.getIdentifier(serialPortName);
+
+            if (serialPortIdentifier == null) {
+                throw new IOException("Unknown serial port");
+            }
+            SerialPort serialPort = serialPortIdentifier.open(this.getClass().getName(), 2000);
+            serialPort.setSerialPortParams(9600, SerialPort.DATABITS_8, SerialPort.STOPBITS_1, SerialPort.PARITY_NONE);
+            serialPort.enableReceiveThreshold(1);
+            serialPort.disableReceiveTimeout();
+
+            InputStream in = serialPort.getInputStream();
+            OutputStream out = serialPort.getOutputStream();
+
+            if (in != null && out != null) {
+                out.flush();
+                if (in.markSupported()) {
+                    in.reset();
+                }
+
+                serialPort.notifyOnDataAvailable(true);
+
+                this.serialPort = serialPort;
+                this.in = in;
+                this.out = out;
+            }
+        } catch (PortInUseException | UnsupportedCommOperationException | IOException e) {
+            throw new EpsonProjectorException(e);
+        }
+    }
+
+    @Override
+    public void disconnect() throws EpsonProjectorException {
+        InputStream in = this.in;
+        OutputStream out = this.out;
+        SerialPort serialPort = this.serialPort;
+
+        if (out != null) {
+            logger.debug("Close serial out stream");
+            try {
+                out.close();
+            } catch (IOException e) {
+                logger.debug("Error occurred when closing serial out stream", e);
```

Comment:
This will log the stack trace. Is logging the exception's message sufficient? Same for below and TCP.
```suggestion
                logger.debug("Error occurred when closing serial out stream: {}", e.getMessage());
```

Reply:
Logging the message now for all.